### PR TITLE
[pyflakes] Avoid false positives in `@no_type_check` contexts (F821, F722)

### PR DESCRIPTION
**Pull request by ntBre**
## Summary

These changes avoid the false positives reported in #13824, where both F821 (undefined name) and F722 (syntax error in forward annotation) were triggered in function signatures decorated with [typing.no_type_check](https://docs.python.org/3/library/typing.html#typing.no_type_check). In line with the discussion on the issue, the code now skips visiting type definitions when in a context where this decorator was found. While the initial report only covered the function decorator, the docs indicate that classes can also be decorated, so this context is also checked when visiting class definitions, and the new tests reflect that.

## Test Plan

These changes were tested by including the code snippets triggering the issue as new snapshot tests. As mentioned above, these snapshots are also augmented with `class` variants. I also added invalid annotations to the function and method bodies to make sure the `no_type_check` context applied to the bodies too.

### CODE REVIEW

File: crates/ruff_linter/src/checkers/ast/mod.rs [hunk 1]
```diff
@@ -1060,6 +1060,17 @@ impl<'a> Visitor<'a> for Checker<'a> {
         self.semantic.flags = flags_snapshot;
     }
 
+    fn visit_decorator(&mut self, decorator: &'a ruff_python_ast::Decorator) {
+        if decorator
+            .expression
+            .as_name_expr()
+            .is_some_and(|name| name.id.as_str() == "no_type_check")
```
[the remark this answers is not in the document]
Comment (charliermarsh) [reply crates/ruff_linter/src/checkers/ast/mod.rs:1067]: So, will this apply to type annotations in the _body_ of the function? And should it, per the spec? (I don't know off-hand.)

**Review by MichaReiser — COMMENTED** [review]
I looked into the `@no_type_check` for red knot and I think we should ignore it for classes, the same as pyright. See https://discuss.python.org/t/no-type-check-decorator/43119. 

Edit: The relevant typing spec change https://github.com/python/typing/pull/1615/files

File: crates/ruff_linter/src/checkers/ast/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ntBre) [reply crates/ruff_linter/src/checkers/ast/mod.rs]: Ah thanks, I missed `match_typing_expr`.

Yes, this should apply to annotations in the body too, based on my reading of the `no_type_check` [docs](https://docs.python.org/3/library/typing.html#typing.no_type_check):
> This works as a class or function [decorator](https://docs.python.org/3/glossary.html#term-decorator). With a class, it applies recursively to all methods and classes defined in that class (but not to methods defined in its superclasses or subclasses). **Type checkers will ignore all annotations in a function or class with this decorator.**

Similarly in the [typing docs](https://typing.readthedocs.io/en/latest/spec/directives.html#no-type-check) linked in what Micha linked:
> If a type checker supports the no_type_check decorator for functions, it should suppress all type errors for the def statement and its body including any nested functions or classes. It should also ignore all parameter and return type annotations and treat the function as if it were unannotated.

I added annotated variables to the test fixtures in my last commit to check this.

**Comment by ntBre** [comment]
Ah okay, happy to revert the class part. That's what I had initially anyway since that's all that was reported in #13824.

**Review by MichaReiser — APPROVED** [review]
Nice. This looks good to me

**Review by charliermarsh — APPROVED** [review]

**Comment by MichaReiser** [comment]
Thanks

**Pull request merged**